File: src/test-trie.c
/*
 * Pucxobot - A robot to play Coup in Esperanto (Puĉo)
 * Copyright (C) 2022  Neil Roberts
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#include "pcx-trie.h"

#include <stdlib.h>
#include <stdio.h>
#include <assert.h>

#include "pcx-buffer.h"
#include "pcx-util.h"

#define MAX_LETTERS_IN_WORD 4
#define N_BITS_PER_LETTER 3

static const char * const
letters[] = {
        "a", "b", "c", "d", "🎩", "Ŝ", "É", "🔚",
};

_Static_assert(PCX_N_ELEMENTS(letters) == (1 << N_BITS_PER_LETTER),
               "There must be one letter string for each possible letter "
               "number");

static int
get_n_words(void)
{
        int n_words_for_length = 1 << N_BITS_PER_LETTER;
        int total_n_words = 0;

        for (int n_letters = 1; n_letters <= MAX_LETTERS_IN_WORD; n_letters++) {
                total_n_words += n_words_for_length;
                n_words_for_length <<= N_BITS_PER_LETTER;
        }

        return total_n_words;
}

static int
get_word_length(uint32_t word_num)
{
        int n_words_for_length = 1 << N_BITS_PER_LETTER;
        int total_n_words = 0;

        for (int n_letters = 1; n_letters <= MAX_LETTERS_IN_WORD; n_letters++) {
                total_n_words += n_words_for_length;

                if (word_num < total_n_words)
                        return n_letters;

                n_words_for_length <<= N_BITS_PER_LETTER;
        }

        assert(!"unknown word_num");
}

static void
get_word(struct pcx_buffer *buf,
         uint32_t word_num)
{
        /* Invert some bits so that the letters aren’t added in order */
        word_num ^= 5;

        pcx_buffer_set_length(buf, 0);

        int length = get_word_length(word_num);

        for (int i = 0; i < length; i++) {
                int letter_num = word_num & ((1 << N_BITS_PER_LETTER) - 1);
                pcx_buffer_append_string(buf, letters[letter_num]);
                word_num >>= N_BITS_PER_LETTER;
        }
}

static bool
check_all_added(struct pcx_trie *trie,
                uint32_t n_words,
                struct pcx_buffer *word_buf)
{
        for (uint32_t word_num = 0; word_num < n_words; word_num++) {
                get_word(word_buf, word_num);

                enum pcx_trie_add_result result =
                        pcx_trie_add_word(trie, (const char *) word_buf->data);

                if (result != PCX_TRIE_ADD_RESULT_ALREADY_ADDED) {
                        fprintf(stderr,
                                "When readding “%s” (%i) result is not "
                                "ALREADY_ADDED\n",
                                (const char *) word_buf->data,
                                word_num);
                        return false;
                }
        }

        return true;
}

int
main(int argc, char **argv)
{
        int ret = EXIT_SUCCESS;

        struct pcx_trie *trie = pcx_trie_new();

        struct pcx_buffer word_buf = PCX_BUFFER_STATIC_INIT;

        int n_words = get_n_words();

        for (uint32_t word_num = 0; word_num < n_words; word_num++) {
                if (!check_all_added(trie, word_num, &word_buf)) {
                        ret = EXIT_FAILURE;
                        break;
                }

                get_word(&word_buf, word_num);

                enum pcx_trie_add_result result =
                        pcx_trie_add_word(trie,
                                          (const char *) word_buf.data);

                if (result != PCX_TRIE_ADD_RESULT_NEW_WORD) {
                        fprintf(stderr,
                                "Added “%s” (%i) but result was not "
                                "PCX_TRIE_ADD_RESULT_NEW_WORD\n",
                                (const char *) word_buf.data,
                                word_num);
                        ret = EXIT_FAILURE;
                        break;
                }
        }

        pcx_buffer_destroy(&word_buf);
        pcx_trie_free(trie);

        return ret;
}

File: src/test-trie.py
#!/usr/bin/python3

# Puxcobot - A robot to play Coup in Esperanto (Puĉo)
# Copyright (C) 2022  Neil Roberts
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import sys
import os.path
import tempfile
import subprocess

WORDS = [
    "terpomo",
    "terpomoj"
    "a",
    "zzz",
    "eĥoŝanĝoĉiuĵaŭde",
]

TEST_WORDS = {
    "terpom" : False,
    "terpomojn" : False,
    "zzzz" : False,
    "-" : False,
    "eĥo" : False,
    "eĥoŝanĝoĉiuĵaŭd🤯" : False,
}

for word in WORDS:
    TEST_WORDS[word] = True

if len(sys.argv) != 3:
    print("usage: test-trie.py <test-trie executable> "
          "<make-dictionary executable>",
          file=sys.stderr)
    sys.exit(1)

test_program = sys.argv[1]
trie_program = sys.argv[2]

with tempfile.NamedTemporaryFile() as dictionary:
    trie_proc = subprocess.Popen([trie_program, dictionary.name],
                                 stdin=subprocess.PIPE)
    trie_proc.communicate("\n".join(WORDS).encode("utf-8"))
    if trie_proc.wait() != 0:
        print("make-dictionary failed", file=sys.stderr)
        sys.exit(1)

    input = []
    expected_output = []

    for word, result in TEST_WORDS.items():
        input.append(word)
        expected_output.append("{}: {}".format(word, "yes" if result else "no"))

    test_proc = subprocess.Popen([test_program, dictionary.name, *input],
                                 encoding="utf-8",
                                 stdout=subprocess.PIPE)

    expected_iter = iter(expected_output)

    result = True

    for line in test_proc.stdout:
        try:
            expected_line = next(expected_iter)
        except StopIteration:
            print("Extra line in output: {}".format(line), file=sys.stderr)
            sys.exit(1)

        if line.rstrip() != expected_line:
            print("Mismatched line:\n"
                  "< {}\n"
                  "> {}".format(expected_line, line), end='')
            result = False

    try:
        next(expected_iter)
    except StopIteration:
        pass
    else:
        print("Not enough lines from test output", file=sys.stderr)
        result = False

    if not result:
        sys.exit(1)

    with tempfile.NamedTemporaryFile(mode="wt", encoding="utf-8") as word_list:
        for word in sorted(WORDS):
            print(word, file=word_list)

        word_list.flush()

        with tempfile.NamedTemporaryFile() as tried_list:
            subprocess.check_call([test_program, dictionary.name],
                                  stdout=tried_list)
            tried_list.flush()

            subprocess.check_call(["diff", word_list.name, tried_list.name])
